src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java

```diff
@@ -30,436 +29,437 @@
+     * @return states that create the same ratio between speeds but scaled down
+     */
+    private static SwerveModuleState[] desaturateSwerveModuleStates(SwerveModuleState[] states) {
+        double desaturationFactor = 1;
+        for (SwerveModuleState state : states) {
+            desaturationFactor = Math.max(desaturationFactor, state.speedMetersPerSecond / ChassisConstants.MAX_VELOCITY);
+        }
+        SwerveModuleState[] desaturatedStates = new SwerveModuleState[states.length];
+        for (int i = 0; i < states.length; i++) {
+            desaturatedStates[i] = new SwerveModuleState(states[i].speedMetersPerSecond / desaturationFactor, states[i].angle);
+        }
+        return desaturatedStates;
+    }
+
+    public ChassisSpeeds getChassisSpeeds() {
+        return kinematics.toChassisSpeeds(getModuleState(Module.FRONT_LEFT),
+                getModuleState(Module.FRONT_RIGHT),
+                getModuleState(Module.BACK_LEFT),
+                getModuleState(Module.BACK_RIGHT));
+    }
+
+    public SwerveModulePosition[] getSwerveModulePositions() {
+        return new SwerveModulePosition[]{
+                frontLeft.getCurrentPosition(),
+                frontRight.getCurrentPosition(),
+                backLeft.getCurrentPosition(),
+                backRight.getCurrentPosition()
+        };
+    }
+
+    public SwerveDriveKinematics getKinematics() {
+        return this.kinematics;
+    }
+
+    public IGyro getGyro() {
+        return gyro;
+    }
+
+    /**
+     * moving a single module by module state
+     */
+    private void setModuleStateForModule(Module module, SwerveModuleState state) {
+        getModule(module).setModuleState(state);
+    }
+
+    /**
+     * for calibration purposes
+     */
+    public void rotateModuleByPower(Module module, double power) {
+        getModule(module).setRotationalPower(power);
+    }
+
+    public void updatePoseEstimationPhotonVision() {
+        poseEstimator.update(getGyroAngle(), getSwerveModulePositions());
+        Photonvision.getInstance().getUpdatedPoseEstimator().ifPresent((EstimatedRobotPose pose) -> poseEstimator.addVisionMeasurement(pose.estimatedPose.toPose2d(), pose.timestampSeconds));
+    }
+
+    public void updatePoseEstimationLimeLight() {
+        boolean poseDifference = odometry.getPoseMeters().getTranslation().getDistance(getRobotPose().getTranslation()) < OdometryConstants.MAX_DISTANCE_TO_FILTER_OUT;
+        boolean shouldUpdateByOdometry = poseDifference && !robotStalling();
+        if (shouldUpdateByOdometry) {
+            poseEstimator.update(getGyroAngle(), getSwerveModulePositions());
+        }
+        if (doVision) {
+            for (Optional<Pair<Pose2d, Double>> target : MultiLimelight.getInstance().getAllEstimates()) {
+                target.ifPresent(this::addVisionMeasurement);
+            }
+        }
+    }
+
+    private boolean moduleStalling(SwerveModule module) {
```

Comment:
stalling is when the motor not moving -> change to another name

Reply:
changed

Reply:
@someone i cant resolve this plz help